### DISCUSSION TITLE
Replace &nbsp; on paste (or any other string/char)

### DISCUSSION THREAD

**Discussion by doits** [ueberdosis/tiptap#1225] [General Questions] [answered]
I'm using latest tiptap 2 beta and wondered if it is possible to replace some characters which are pasted into the editor, in my case replace all `&nbsp;` (or `\u00A0`) with a regular space.

I looked into paste rules in other extensions and tried to create a new extension, but I'm at a dead end here:

```js
import { Extension, markPasteRule } from '@tiptap/core'

const nbspReplacer = Extension.create({
  addPasteRules() {
    return [
      markPasteRule(/\u00A0/, this.type)
    ]
  }
})
```

I don't see a way to specify how to replace it, e.g. I don't want to create a new node or something but only replace single chars. `markPasteRule` might be the wrong tool for it.

https://www.tiptap.dev/examples/savvy does exactly this with *input* rules (replacing something with emojis) ...

```js
import { Extension } from '@tiptap/core'
import { InputRule } from 'prosemirror-inputrules'

export const SmilieReplacer = Extension.create({
  name: 'smilieReplacer',

  addInputRules() {
    return [
      new InputRule(/-___- /, '😑 '),
      [...]
    ]
  },
})
```

... but I need it only for *paste* rules.

Is this possible?

**Comment by cadars** [comment]
Did you succeed? I'm trying the exact same thing, with no success.

**Reply by doits** [reply]
Unfortunately I don't have a solution yet.

**Comment by cadars** [comment]
Would this could be done by using the newly introduced `textPasteRule`?

**Comment by rfgamaral** [comment, marked as answer]
@doits What about this?

```ts
const NBSPReplacer = Extension.create({
    name: 'npspReplacer',
    addPasteRules() {
        return [
            textPasteRule({ find: /&nbsp;/g, replace: ' ' }),
            textPasteRule({ find: /\\u00A0/g, replace: ' ' }),
        ]
    },
})
```

**Reply by cadars** [reply]
Thank you!! ☺︎ ☺︎ ☺︎ (There is one extra backslash in your regex though, so here is it again)

```js
import { Extension, textPasteRule } from '@tiptap/core';

const ReplaceNbsp = Extension.create({
  name: 'ReplaceNbsp',
  addPasteRules() {
    return [
      textPasteRule({ find: /&nbsp;/g, replace: ' ' }),
      textPasteRule({ find: /\u00A0/g, replace: ' ' }),
    ]
  },
})
```

**Reply by rfgamaral** [reply]
The extra backslash was on purpose because I wasn't sure what did you want to catch, the actual unicode string or the unicode character.

But catching that unicode character didn't work for me, but maybe I didn't test it right.

**Reply by philippkuehn** [reply]
Not sure but maybe you need the `u` flag for that?

**Reply by cadars** [reply]
> But catching that unicode character didn't work for me

Looks like the replacement doesn't work when it's pasted as the first node in the document? (I need to test without the placeholder extension just in case)

**Reply by rfgamaral** [reply]
@cadars It worked for me.

**Reply by cadars** [reply]
Could you try copy and pasting the post titles of https://kottke.org/ ? In this case the `&nbsp;` is inside a link, inside a heading.

**Reply by rfgamaral** [reply]
I'm sorry, but I don't see `&nbsp;` anywhere in that page.

**Reply by cadars** [reply]
Sorry, better here: https://codesandbox.io/s/musing-moon-b38t9?file=/src/App.js

nbsps are replaced by `X`:
- copy the paragraph and paste multiple times, it should work
- empty the editor and do the same with the heading, the first one keeps the nbsps

**Reply by philippkuehn** [reply]
oh no, seems like pasterules doesn’t work at all in this situation.

**Reply by philippkuehn** [reply]
should be fixed with this patch: https://github.com/ueberdosis/tiptap/commit/ff67ee1da380d8308e85fa4b0386ea6947ec7ff1

**Reply by cadars** [reply]
Thanks much! One last question though 😅, in the codesandbox above, if you change `'X'` for `' '`, it does not perform the replacement: what would be the correct way to specify the regular space character?

**Reply by rfgamaral** [reply]
I'm not sure what's happening, but I don't seem to have any issues with the demo above (which I don't think is using @philippkuehn patch):

https://user-images.githubusercontent.com/96476/138442702-1cc560df-69a8-4baa-afb0-a2bfdccbebd1.mp4

**Reply by cadars** [reply]
You're right, my bad, I forgot that the .ProseMirror container doesn’t collapse multiple spaces by default.
Thanks for your time!
